ZOOKEEPER-3156: Add in option to canonicalize host name

revans2 opened the pull request.

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -990,6 +992,27 @@ private void sendPing() {
+                //ignored ...
+            }
+
+            if (canonicalize) {
```
You only want to do this if SASL client is enabled. Hence the name of the switch contains `sasl.client`.

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -990,6 +992,27 @@ private void sendPing() {
+            if (canonicalize) {
+                try {
+                    InetAddress ia = addr.getAddress();
+                    LOG.warn("ia {}", ia);
```
Debugging?

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -990,6 +992,27 @@ private void sendPing() {
+                    InetAddress ia = addr.getAddress();
+                    LOG.warn("ia {}", ia);
+                    if (ia == null) {
+                        ia = InetAddress.getByName(addr.getHostName());
```
If the original address is unresolved, you should throw exception immediately. That means HostProvider already tried to resolve the address, but failed. No point trying it again here.

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -990,6 +992,27 @@ private void sendPing() {
+                        ia = InetAddress.getByName(addr.getHostName());
+                    }
+                    String host = (ia != null) ? ia.getCanonicalHostName() : addr.getHostName();
+                    addr = new InetSocketAddress(InetAddress.getByAddress(host, ia.getAddress()), addr.getPort());
```
I'm thinking of how much value does it have to replace the address with the canonicalized version for the entire client. We might want to implement this strictly and only for `ZooKeeperSaslClient`.

I'd rather move this logic to the creation of `ZooKeeperSaslClient`. 

Something like:
```java
zooKeeperSaslClient =  new ZooKeeperSaslClient(principalUserName + "/" + canonicalize ? addr.getAddress().getCanonicalHostName() : addr.getHostName());
```

revans2 replied:
I put it here because there is a possible race.  The entire reason we have the setup we do is so that we can change the nodes in the cluster without changing the config on the client side.  If the code that establishes the connection uses a different address from the code that creates the principal there is the possibility, because of the magic of DNS caching, that the connection would be made to a different box from the one the SASL client is expecting.  If that happens, and we have the default Client section in the jaas conf, it will not result in an error.  Instead the ZK client logs something about SASL failed and still connects but is not authorized to do anything.  If it failed and retried a different node I would be happy to move it, but it does not and that failure would not be transparent to the end user.

revans2 replied:
But you said that we already have the IP address, which is what it is going to use, so never mind, it should be fine.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2282/

revans2 commented:
@anmolnar I think I have addressed all of your comments.  Please have another look.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2284/

eolivelli commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -997,12 +999,31 @@ private void startConnect(InetSocketAddress addr) throws IOException {
+                    }
+                    //Update the actual address so we are
+                    hostName = ia.getCanonicalHostName();
+                    LOG.debug("Canonicalized address to {}", hostName);
```
Nit: if LOG.isDebugEnabled

anmolnar replied:
+1

anmolnar commented:
@eolivelli I like your idea about the unit test.

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -997,12 +999,31 @@ private void startConnect(InetSocketAddress addr) throws IOException {
+                if (canonicalize) {
+                    InetAddress ia = addr.getAddress();
+                    if (ia == null) {
+                        throw new IllegalArgumentException("Connection address should have already been resolved by the HostProvider.");
```
Would you please change this to something like 'Unable to canonicalize address {}, because it's not resolveable'?

anmolnar commented on src/java/main/org/apache/zookeeper/ClientCnxn.java:
```diff
@@ -997,12 +999,31 @@ private void startConnect(InetSocketAddress addr) throws IOException {
+                        throw new IllegalArgumentException("Connection address should have already been resolved by the HostProvider.");
+                    }
+                    //Update the actual address so we are
+                    hostName = ia.getCanonicalHostName();
```
You might want to do the following:
```java
String canonicalHostName = ia.getCanonicalHostName();
if (!canonicalHostName.equals(ia.getHostAddress())) {
    hostName = canonicalHostName;
}
```

In order to avoid using literal IP address when security check fails.

revans2 commented:
I just addressed the review comments if things look good I am happy to squash commits.  I'll also put up pull requests to the other lines, now that it looks like we have the majority of the code worked out.

eolivelli commented:
Overall looks good to me

What about adding a test case ?

revans2 commented:
So I am running into some issues with the tests as all of the methods to InetSocketAddress are marked as final and as such I cannot mock them. This means I would need a set of host names that are real, and accessible from any build host you would want to run the test on and never going to be taken down, or I would need to create another class that wraps InetSocketAddress for the canonicalization that would allow me to do the mocking at that level.  Not sure which you would prefer.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2299/

eolivelli commented:
Do we have PowerMockito on ZK classpath for tests ?

revans2 commented:
I tried to add in powermock too, but it conflicts with mockito (power mock uses a much older version of mockito-core).  I'll try and use a different API for powermock instead of mockito.

revans2 commented:
I just added in a test and made the code match closer to how master/3.5 work for the canonical host name.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2302/

revans2 commented:
There were some comments in the other pull request about the test code.  I will address them in both places.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2326/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2323/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2322/

revans2 commented:
Thanks for all of the reviews I just rebased and squashed commits.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2347/

anmolnar commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2469/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2551/

anmolnar commented:
Committed. Thanks @revans2 !
Please close this PR.

revans2 commented:
Thanks
